### PR TITLE
Add refs/decomps for dot/vdot

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108194
* #108188

Follow-up on https://github.com/pytorch/pytorch/issues/108127#issuecomment-1698142427

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1168,15 +1168,10 @@ def make_arg_conj(size):
...
 
-    yield ErrorInput(SampleInput(make_input(1), args=(make_input(3, dtype=torch.float16),)),
-                     error_regex='dot : expected both vectors to have same dtype')
     yield ErrorInput(SampleInput(make_input(1, 1), args=(make_input(3),)),
                      error_regex='1D tensors expected')
     yield ErrorInput(SampleInput(make_input(9), args=(make_input(3),)),
                      error_regex='inconsistent tensor size')
-    if device != "cpu":
-        ErrorInput(SampleInput(make_input(3), args=(make_input(3, device="cpu"),)),
-                   error_regex='Expected all tensors to be on the same device')
```
Comment (peterbell10) [inline torch/testing/_internal/common_methods_invocations.py:1179]: Why are we changing the tests here? Shouldn't you fix the refs to match the test.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -20957,6 +20952,26 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
         "_refs.vsplit",
         torch_opinfo_name="vsplit",
     ),
+    PythonRefInfo(
+        "_refs.dot",
+        torch_opinfo_name="dot",
+        # .conj() returns a view on complex tensors??
+        validate_view_consistency=False,
```
Comment (peterbell10) [inline torch/testing/_internal/common_methods_invocations.py:20959]: Why aren't we getting the view behavior right here?

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:20959]: At a guess, `return torch.dot(self, other.conj()).conj()` probably needs to come outside of `elementwise_type_promotion_wrapper`.

Comment (lezcano) [reply torch/testing/_internal/common_methods_invocations.py:20959]: Not quite. I checked under this branch:
```python
    elif other.is_conj():
        return torch.dot(self, other.conj()).conj()
```
which is the one that triggers the issue, and `torch.dot(self, other.conj())` is not a view, but `torch.dot(self, other.conj()).conj()` is a view.
On the other hand, if I put the same code in ATen, it says that `torch.dot(self, other.conj()).conj()` is conj but not a view... I'm submitting a different fix.

Comment (lezcano) [reply torch/testing/_internal/common_methods_invocations.py:20959]: Nvm, the other fix doesn't work. I'm not quite sure what's going on there.
I'm voting for leaving as is right now as we don't even support complex numbers at the moment in inductor.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:1179]: Adding more functionality to the ref seems fine, but removing testing from eager definitely seems bad,

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:20959]: Okay the eager result having `is_conj()` but not `_is_view()` is weird but not really our fault here, so I guess thats fine. This comment should be updated to reflect that though; `.conj()` is expected to return a view so the comment doesn't make much sense.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (peterbell10) [reply torch/testing/_internal/common_methods_invocations.py:1179]: Could you override the `error_inputs_func` of the ref info to pass in an `is_ref=True` option?

Comment (lezcano) [reply torch/testing/_internal/common_methods_invocations.py:1179]: Fair enough. Done.

**Review by peterbell10 — APPROVED** [review]

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 3]
```diff
@@ -496,12 +496,6 @@ inline void dot_check(const Tensor& self, const Tensor& other) {
       " and ",
       other.numel(),
       " elements respectively");
-  TORCH_CHECK(
-      self.device() == other.device(),
-      "expected all tensors to be on the same device. Found: ",
-      self.device(),
-      ", ",
-      other.device());
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Blas.cpp:504]: @peterbell10 removing this as it's unreachable. We currently check this in the wrapper and throw
```
Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument tensor in method wrapper_CUDA__dot
```

**Pull request closed without merging**